### PR TITLE
[Enchancement](optimize) set result_size_hint to filter_block

### PR DESCRIPTION
**Pull request by BiteTheDDDDt**
# Proposed changes

```sql
SELECT SUM(LO_REVENUE - LO_SUPPLYCOST)
FROM lineorder_flat
WHERE
    C_REGION = 'AMERICA'
    AND S_REGION = 'AMERICA'
    AND( P_MFGR = 'MFGR#1'
    OR P_MFGR = 'MFGR#2');

```
before:
图片
after:
图片
## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/columns/columns_common.cpp [hunk 1]
```diff
@@ -106,14 +106,14 @@ struct ResultOffsetsBuilder {
...
-        res_offsets.push_back(current_src_offset);
+        res_offsets.push_back_without_reserve(current_src_offset);
     }
 
     template <size_t SIMD_BYTES>
     void insertChunk(const IColumn::Offset* src_offsets_pos, bool first,
                      IColumn::Offset chunk_offset, size_t chunk_size) {
         const auto offsets_size_old = res_offsets.size();
-        res_offsets.resize(offsets_size_old + SIMD_BYTES);
+        res_offsets.resize_assume_reserved(offsets_size_old + SIMD_BYTES);
```
Comment (spaces-X) [inline be/src/vec/columns/columns_common.cpp:116]: Make sure that `reserve()` function has been called before here in all cases.

File: be/src/vec/core/block.cpp [hunk 2]
```diff
@@ -638,7 +624,7 @@ void filter_block_internal(Block* block, const IColumn::Filter& filter, uint32_t
         if (count != block->rows()) {
             for (size_t i = 0; i < column_to_keep; ++i) {
                 block->get_by_position(i).column =
-                        block->get_by_position(i).column->filter(filter, 0);
+                        block->get_by_position(i).column->filter(filter, count);
```
Comment (spaces-X) [inline be/src/vec/core/block.cpp:627]: Please add UT for the new function call

File: be/src/vec/columns/columns_common.cpp [hunk 3]
```diff
@@ -106,14 +106,14 @@ struct ResultOffsetsBuilder {
 
     void insertOne(size_t array_size) {
```
Comment (HappenLee) [inline be/src/vec/columns/columns_common.cpp:107]: change `insert_one`

File: be/src/vec/columns/columns_common.cpp [hunk 4]
```diff
@@ -106,14 +106,14 @@ struct ResultOffsetsBuilder {
...
         current_src_offset += array_size;
-        res_offsets.push_back(current_src_offset);
+        res_offsets.push_back_without_reserve(current_src_offset);
     }
 
     template <size_t SIMD_BYTES>
     void insertChunk(const IColumn::Offset* src_offsets_pos, bool first,
```
Comment (HappenLee) [inline be/src/vec/columns/columns_common.cpp:113]: same as up

File: be/src/vec/columns/columns_common.cpp [hunk 5]
```diff
@@ -151,15 +151,15 @@ void filter_arrays_impl_generic(const PaddedPODArray<T>& src_elems,
         LOG(FATAL) << "Size of filter doesn't match size of column.";
     }
 
+    constexpr int assume_string_length = 5;
```
Comment (HappenLee) [inline be/src/vec/columns/columns_common.cpp:154]: ASSUME_STRING_LENGTH

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**